### PR TITLE
[ADD] coal_petroleum: module for fossil fuels

### PR DESCRIPTION
**Pull request by paso-odoo**
This module contains the data of

- Automated action, Window action, Attachments, Access rules, Knowledge article, Knowledge cover, Knowledge article member, Product, Project template, Product category, Product attribute, Quality point, Worksheet template.

and Demo data as
- Product supplier info, Partners.

task: 3460423

### CODE REVIEW

**Review by jva-odoo — COMMENTED** [review]
don't create multi-line if possible

File: industry_coal_petroleum/data/knowledge_article.xml
Comment (jva-odoo) [inline industry_coal_petroleum/data/knowledge_article.xml:2]: article is updatable so remove `noupdate`

File: industry_coal_petroleum/data/product_template.xml
Comment (jva-odoo) [inline industry_coal_petroleum/data/product_template.xml:6]: by default, it's `manual` so remove it

File: industry_coal_petroleum/data/product_template_attribute_line.xml
Comment (jva-odoo) [inline industry_coal_petroleum/data/product_template_attribute_line.xml:7]: it's auto compute so remove it

File: industry_coal_petroleum/data/stock_warehouse.xml
Comment (jva-odoo) [inline industry_coal_petroleum/data/stock_warehouse.xml:3]: update it normally

File: industry_coal_petroleum/demo/purchase_order.xml
Comment (jva-odoo) [inline industry_coal_petroleum/demo/purchase_order.xml:30]: remove this

File: industry_coal_petroleum/data/product_product.xml
Comment (jva-odoo) [inline industry_coal_petroleum/data/product_product.xml:3]: remove this hash from all records
```suggestion
    <record id="product_product_10" model="product.product">
```

File: industry_coal_petroleum/demo/purchase_order_post.xml
Comment (jva-odoo) [inline industry_coal_petroleum/demo/purchase_order_post.xml:66]:
```suggestion
```
not needed it's auto call from create

File: industry_coal_petroleum/demo/res_partner.xml
Comment (jva-odoo) [inline industry_coal_petroleum/demo/res_partner.xml:10]:
```suggestion
```

File: industry_coal_petroleum/demo/sale_order_line.xml
Comment (jva-odoo) [inline industry_coal_petroleum/demo/sale_order_line.xml:7]:
```suggestion
```

File: industry_coal_petroleum/demo/sale_order_post.xml
Comment (jva-odoo) [inline industry_coal_petroleum/demo/sale_order_post.xml:24]:
```suggestion
```

**Review by jva-odoo — CHANGES_REQUESTED** [review]
all other things it's good just change this and we good to go

File: industry_coal_petroleum/demo/purchase_order_post.xml
Comment (jva-odoo) [reply industry_coal_petroleum/demo/purchase_order_post.xml:66]: my bad

**Comment by paso-odoo** [comment]
@jva-odoo 

Tested in 17.0

**Review by jva-odoo — APPROVED** [review]

**Comment by jva-odoo** [comment]
squashed and rebase with 17.0

**Comment by jva-odoo** [comment]
runbot is red because of wkhtmltopdf warring

File: coal_petroleum/__manifest__.py
Comment (paso-odoo) [reply coal_petroleum/__manifest__.py:14]: This module installs the dependency of accounting modules as per the main DB 
i.e. https://trading-coal.odoo.com/

File: coal_petroleum/data/ir_rule.xml
Comment (paso-odoo) [reply coal_petroleum/data/ir_rule.xml:1]: No

File: coal_petroleum/data/knowledge_article_member.xml
Comment (paso-odoo) [reply coal_petroleum/data/knowledge_article_member.xml:3]: Please check this line
`<field name="article_id" ref="coal_petroleum.knowledge_article_4"/>`

Here If we don't use `knowledge_article_member1` then the main user cannot see the article in sidebar.

File: coal_petroleum/demo/sale_order_post.xml
Comment (paso-odoo) [reply coal_petroleum/demo/sale_order_post.xml:78]: @vava-odoo 

No, we shouldn't perform grouping in this scenario. 

In both Sales Orders, the customer is the same. If we group them and invoke the _create_invoices function, it will generate a single invoice incorporating all the products from both Sales Orders. 

If you have any suggestions, please let me know.

File: coal_petroleum/data/product_product.xml
Comment (paso-odoo) [reply coal_petroleum/data/product_product.xml:2]: Yes, All the products are connected to this particular industry.
So, It is added to the data rather than the demo file.

File: coal_petroleum/data/res_config_settings.xml
Comment (paso-odoo) [reply coal_petroleum/data/res_config_settings.xml:2]: We should use data noupdate="1" because of the function here

File: coal_petroleum/demo/purchase_order_line.xml
Comment (paso-odoo) [reply coal_petroleum/demo/purchase_order_line.xml:15]: Not even in the main db

File: coal_petroleum/demo/purchase_order_post.xml
Comment (paso-odoo) [reply coal_petroleum/demo/purchase_order_post.xml:187]: @vava-odoo 

It's not a duplicate code.

Here we use **2-Step Delivery**, So there are 2 Receipts for 1 Purchase Order, We must need to do validate one by one.

If you have any other standard solution, please let me know.

Thanks in advance!

File: coal_petroleum/data/knowledge_article_member.xml
Comment (vava-odoo) [reply coal_petroleum/data/knowledge_article_member.xml:3]: Really? (not really familiar with that App) 
I thought the following lines were sufficient? (on knowledge.article)
```xml
        <field name="internal_permission">write</field>
        <field name="is_article_visible_by_everyone" eval="True"/>
```

File: coal_petroleum/data/res_config_settings.xml
Comment (vava-odoo) [reply coal_petroleum/data/res_config_settings.xml:2]: Ok, then enclose the record with `<data noupdate="1">`. We don't want to re-activate settings if they unactivate them.
```suggestion
<odoo>
    <data noupdate="1">
        <record id="res_config_settings_multi_step_routes" model="res.config.settings">
            ...
        </record>
    </data>
    
    <function model="res.config.settings" name="execute">
        ...
```

File: coal_petroleum/demo/purchase_order_line.xml
Comment (vava-odoo) [reply coal_petroleum/demo/purchase_order_line.xml:15]: Ok, so 0 tolerance on this one... Doesn't make sense to me. Would be worth checking with the person who created the db.

File: coal_petroleum/demo/purchase_order_post.xml
Comment (vava-odoo) [reply coal_petroleum/demo/purchase_order_post.xml:187]: I see that it is done like that in other tests in standard. Maybe @pimodoo has a better idea, otherwise let's leave it like that.

File: coal_petroleum/data/res_config_settings.xml
Comment (vava-odoo) [reply coal_petroleum/data/res_config_settings.xml:2]: In fact, all industries have a res_config_settings starting with `<odoo noupdate="1">`. There is no point in calling the function at the update, since no modification is done to the settings.

File: coal_petroleum/demo/purchase_order_post.xml
Comment (vava-odoo) [reply coal_petroleum/demo/purchase_order_post.xml:187]: Just discussed it with @pimodoo. It is right, you need to validate twice.
In the current version, you check the state of the pickings, and those will automatically change after the first validation. However I'd find it clearer if we distinguish them within the search, i.e. first one with a `picking_type_code='internal'` and second one with a `picking_type_code='outgoing'`.

File: coal_petroleum/data/knowledge_article.xml
Comment (vava-odoo) [inline coal_petroleum/data/knowledge_article.xml:2]:
```suggestion
<odoo>
```

File: coal_petroleum/data/res_config_settings.xml
Comment (vava-odoo) [inline coal_petroleum/data/res_config_settings.xml:9]: Is it really useful here?

File: coal_petroleum/demo/purchase_order_post.xml
Comment (vava-odoo) [reply coal_petroleum/demo/purchase_order_post.xml:187]: last comment was not taken into account

File: coal_petroleum/demo/sale_order_line.xml
Comment (vava-odoo) [inline coal_petroleum/demo/sale_order_line.xml:4]: name of SOL defaults to name of product
```suggestion
```

**Review by vava-odoo — CHANGES_REQUESTED** [review]
Thanks for the changes! Almost there!

**Comment by paso-odoo** [comment]
@vava-odoo 
The requested changes are done!
Thanks!

File: coal_petroleum/data/product_category.xml
Comment (vava-odoo) [inline coal_petroleum/data/product_category.xml:3]: Shouldn't the product category be a child of "All" ?

**Review by vava-odoo — COMMENTED** [review]
Thanks @paso-odoo 
Last comments before merging this.

File: coal_petroleum/data/ir_model_fields.xml
Comment (paso-odoo) [reply coal_petroleum/data/ir_model_fields.xml:136]: No, All 3 fields are required.

File: coal_petroleum/data/ir_model_fields.xml
Comment (paso-odoo) [reply coal_petroleum/data/ir_model_fields.xml:154]: No, Vehicle is not mandatory for every transfer

**Comment by paso-odoo** [comment]
@vava-odoo 

The requested changes are done!

File: coal_petroleum/data/ir_model_fields.xml
Comment (vava-odoo) [reply coal_petroleum/data/ir_model_fields.xml]: But then the knowledge needs to be updated as it mentions 

> It is required to provide the vehicle number for the purpose of formal documentation

Comment (paso-odoo) [reply coal_petroleum/data/ir_model_fields.xml]: As per task owner, 
If I make this field mandatory then it will be visible in all transfers. In case of internal transfer there will be no vehicle. 

We need to put attrs in view or domain in the field and make this field mandatory.

Comment (paso-odoo) [reply coal_petroleum/data/ir_model_fields.xml]: @vava-odoo I have updated the code based on the conversation.

Thank you!

File: coal_petroleum/data/ir_model_access.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_model_access.xml:12]: Isn't it induced by the one defined on user, just below?
```suggestion
```

File: coal_petroleum/data/ir_model_access.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_model_access.xml:32]: Same 
```suggestion
```

File: coal_petroleum/data/ir_ui_view.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_ui_view.xml:150]: Useless string
```suggestion
                            <field name="x_net_calorific_value"/>
```

File: coal_petroleum/data/ir_ui_view.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_ui_view.xml:147]: I don't get what this field is used for. Could you explain?

File: coal_petroleum/data/ir_ui_view.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_ui_view.xml:149]: Same comment

File: coal_petroleum/data/ir_ui_view.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_ui_view.xml:175]: Could you improve/simplify the xpath (to ease maintenance)?

File: coal_petroleum/data/ir_ui_view.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_ui_view.xml:195]: Same

File: coal_petroleum/data/ir_ui_view.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_ui_view.xml:211]:
```suggestion
                    <h1 invisible="context.get('default_x_quality_check_id')">
```

**Review by vava-odoo — COMMENTED** [review]
A few more, waiting for functional approval

**Comment by paso-odoo** [comment]
@vava-odoo 

Done!

File: coal_petroleum/data/ir_model_fields.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_model_fields.xml:154]: This field is now useless
```suggestion
```

File: coal_petroleum/data/ir_model_fields.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_model_fields.xml:186]: Same
```suggestion
```

File: coal_petroleum/data/ir_model_fields.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_model_fields.xml:161]: Same
```suggestion
```

File: coal_petroleum/data/ir_model_fields.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_model_fields.xml:193]: Same
```suggestion
```

**Review by vava-odoo — APPROVED** [review]
Ok, just a few fields that became useless since your last changes and should be removed, and we're good to go (if PO approves). Thanks @paso-odoo!

File: coal_petroleum/data/ir_actions_server.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_actions_server.xml:23]: We should also avoid negative values
```suggestion
                if not (0. <= record.x_po_tolerance_1 < 1.0):
```

File: coal_petroleum/data/ir_actions_server.xml
Comment (vava-odoo) [inline coal_petroleum/data/ir_actions_server.xml:35]:
```suggestion
                if not (0. <= record.x_tolerance < 1.0):
```

**Review by vava-odoo — APPROVED** [review]
Hi @paso-odoo and @shag-odoo 
Thanks for the update, seems in line with what was requested by the PO.
Just a small remark and we merge.
Cheers

**Comment by paso-odoo** [comment]
@vava-odoo 

Done!
We can move further!

Thanks

**Pull request merged**
